packages/next-server/lib/head.tsx

```diff
@@ -137,32 +137,62 @@ function reduceComponents(
     })
 }
 
-const Effect = withSideEffect()
-
 /**
  * This component injects elements to `<head>` of your page.
  * To avoid duplicated `tags` in `<head>` you can use the `key` property, which will make sure every tag is only rendered once.
  */
-function Head({ children }: { children: React.ReactNode }) {
-  return (
-    <AmpStateContext.Consumer>
-      {ampState => (
-        <HeadManagerContext.Consumer>
-          {updateHead => (
-            <Effect
-              reduceComponentsToState={reduceComponents}
-              handleStateChange={updateHead}
-              inAmpMode={isInAmpMode(ampState)}
-            >
-              {children}
-            </Effect>
-          )}
-        </HeadManagerContext.Consumer>
-      )}
-    </AmpStateContext.Consumer>
-  )
+function Head({ children }: { children: JSX.Element }) {
+  const ampState = React.useContext(AmpStateContext)
+  const updateHead = React.useContext(HeadManagerContext)
+  const instanceRef = React.useRef(children)
+
+  // Update the instanceRef every render
+  instanceRef.current = children
```

Comment:
Not 100% sure but I think that mutating the ref in render will be "unsafe" in concurrent mode

Reply:
Very good catch, mutation must happen in useEffect.

Reply:
While this is very true for operations that could get out of sync between render-attempts and actual reconciler commits, eg. `ref.current++`, this shouldn't be an issue here as it's merely a synchronous way of keeping track of the latest `children` that have been passed to the component, and has no dependency on the previous value of the ref.

Reply:
In concurrent mode a render can be thrown away and thus not reflective of the actual state. 🤔

Reply:
If a previous render does get thrown away, we can still trust that the value will get updated again in a future render before the final commit. It's also valuable to look at how the ref is being used to know if it's safe or not to do this:
- `instanceRef` is being used synchronously in the render function during SSR below. While not technically possible yet, concurrent mode will likely become possible during SSR very soon, so we need to be mindful of that `typeof window === 'undefined'` code block. Luckily, the `instanceRef` we're using here is being added to a `Set` in that block (which will get deduped during multiple non-commit renders) and `emitUpdate` is immutably constructing the new head list regardless of the previous call to `emitUpdate`. With that said, it would still be safe that multiple renders happens before an actual commit on the server, even with a hypothetical concurrent mode.
- For the code blocks using `instanceRef` on the client, both blocks are wrapped in `useEffect`, so we're also safe that the value is the most up to date version of the `instanceRef` and that it will only get executed during the commit phase.

I can see that although this code is safe, that may not be very clear to most working on it, so we can talk about alternatives. One simple alternative could be adding the `instanceRef` update logic to each `useEffect` block that relies on it and also the `typeof window === 'undefined'` block that uses it in SSR. eg.

```diff
function Head({ children }: { children: JSX.Element }) {
  const ampState = React.useContext(AmpStateContext)
  const updateHead = React.useContext(HeadManagerContext)
  const instanceRef = React.useRef(children)

- // Update the instanceRef every render
- instanceRef.current = children

  const inAmpMode = isInAmpMode(ampState)

  const emitUpdate = () => {
    headState = reduceHeadInstances([...headInstanceRefs], {
      inAmpMode,
    })
    if (updateHead) {
      updateHead(headState)
    }
  }

  // The `useEffect` hook is only called on the client.
  // This means we need to simulate the component being mounted and trust that
  // `rewind()` is invoked to clean up these entries.
  if (typeof window === 'undefined') {
+   // Update the instanceRef
+   instanceRef.current = children
    headInstanceRefs.add(instanceRef)
    emitUpdate()
  }

  // We need to register this head instance on mount and unregister it when
  // unmounted.
  React.useEffect(() => {
+   // Update the instanceRef
+   instanceRef.current = children
    headInstanceRefs.add(instanceRef)
    // n.b. emitUpdate() for mount is called in below effect
    return () => {
      headInstanceRefs.delete(instanceRef)
      emitUpdate()
    }
  }, [])

  // Trigger an update on mount and every update (note the undefined dependency
  // array)
  React.useEffect(() => {
    emitUpdate()
  })

  return null
}
```

Reply:
This suggestion incorrectly puts the `instanceRef.current` update in the second `Effect` (which only adds the object to the `Set`) instead of the third `Effect`, which is what actually would cause the `Set` to be read and needs to be updated on every render; but putting the assignments in the server-side block and the `Effect` block that calls `emitUpdate()` look sufficient.

The complication here is that different `<Head/>` components might commit at different times (depending on `Suspense` boundaries), but they all call `emitUpdate`; if multiple `Head` instances participate in a render pass, but one of them is inside a boundary that suspends, the `emitUpdate`s called by the `Head`s inside the subtree that did commit would see the `render`-updated `instanceRef.current`.

There probably is room here for optimizing updates to only handle each `Head`'s own `children`; the tricky part is SSR.